src/track/beats.cpp

```diff
@@ -734,6 +734,28 @@ int Beats::numBeatsInRange(audio::FramePos startPosition, audio::FramePos endPos
     return i - 2;
 };
 
+double Beats::numFractionalBeatsInRange(audio::FramePos startPos,
+        audio::FramePos endPos) const {
+    // find first beat in range (pos >= startPos)
+    audio::FramePos firstBeatPos = findNextBeat(startPos);
+
+    // maybe shift range to start at first beat
+    audio::FrameDiff_t firstBeatOffset = firstBeatPos - startPos;
+    endPos += firstBeatOffset;
```

Comment:
This only works if the beats have an equal length. For details around a position you may use: 
`BpmControl::getBeatContext()`. This way you get the beat fraction from before and after the fist and last beat. 
We may move this function to this class?

Reply:
That function is helpful, thanks!
Yes, that (and `BpmControl::getBeatContextNoLookup`) should be in the `Beats` class.

Anyhow, the current implementation works well with constant BPM.
For tracks with variable BPM, the fractional size is just an estimation. Though IMO that doesn't matter since with var. BPM tracks `loop_halve/_double` aren't working as expected anyway AFAICT: the loop is simply scaled based on loop_in position and number of beats, i.e. it doesn't consider shorter/longer beats inside or after the loop.

Reply:
Ok I understand. 

What could be a real live use case? 
A beat loop with an integer number moved by a half beat and a tempo change inside a loop. 

Which beat length is used for moving? Probably the first beat? After moving, the loop will start at the middle of the fist beat and end somewhere else in the last beat. 
I think the loop size should be still without a fraction, right? 

What happens if the loop is moved fully in the new tempo region? 

We need to test this.

Reply:
Loops that contain tempo changes are moved correctly: both loop_in and loop_out are moved to the correct fractional position. This works with all loop sizes. :+1:
I tested with 4 and 3.5 beats loop moved by 1/2 and 8 beats several times while the tempo changed from ~110 to ~140.